File: apps/ruby/test/fizz_buzz/domain/type/fizz_buzz_type_test.rb
# frozen_string_literal: true

require_relative '../../../test_helper'
require_relative '../../../../lib/fizz_buzz/fizz_buzz'

class FizzBuzzTypeTest < Minitest::Test
  describe 'タイプ1の場合' do
    def setup
      @type = FizzBuzzType.create(FizzBuzzType::TYPE_01)
    end

    def test_1を渡したら文字列1を返す
      assert_equal '1', @type.generate(1).to_s
    end

    def test_3を渡したらFizzを返す
      assert_equal 'Fizz', @type.generate(3).to_s
    end

    def test_5を渡したらBuzzを返す
      assert_equal 'Buzz', @type.generate(5).to_s
    end

    def test_15を渡したらFizzBuzzを返す
      assert_equal 'FizzBuzz', @type.generate(15).to_s
    end
  end

  describe 'タイプ2の場合' do
    def setup
      @type = FizzBuzzType.create(FizzBuzzType::TYPE_02)
    end

    def test_3を渡したら文字列3を返す
      assert_equal '3', @type.generate(3).to_s
    end

    def test_15を渡したら文字列15を返す
      assert_equal '15', @type.generate(15).to_s
    end
  end

  describe 'タイプ3の場合' do
    def setup
      @type = FizzBuzzType.create(FizzBuzzType::TYPE_03)
    end

    def test_3を渡したら文字列3を返す
      assert_equal '3', @type.generate(3).to_s
    end

    def test_15を渡したらFizzBuzzを返す
      assert_equal 'FizzBuzz', @type.generate(15).to_s
    end
  end

  describe 'ファクトリメソッド' do
    def test_TYPE_01を指定するとFizzBuzzType01が返る
      type = FizzBuzzType.create(FizzBuzzType::TYPE_01)
      assert_instance_of FizzBuzzType01, type
    end

    def test_未定義のタイプを指定するとエラーが発生する
      assert_raises RuntimeError do
        FizzBuzzType.create(99)
      end
    end
  end
end

File: apps/kotlin/src/test/kotlin/fizzbuzz/FizzBuzzTypeTest.kt
package fizzbuzz

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class FizzBuzzTypeTest {
    @Test fun `TYPE_01 は通常の FizzBuzz`() {
        assertEquals("1", FizzBuzzType.TYPE_01.generate(1))
        assertEquals("Fizz", FizzBuzzType.TYPE_01.generate(3))
        assertEquals("Buzz", FizzBuzzType.TYPE_01.generate(5))
        assertEquals("FizzBuzz", FizzBuzzType.TYPE_01.generate(15))
    }

    @Test fun `TYPE_02 は数字のみ`() = assertEquals("3", FizzBuzzType.TYPE_02.generate(3))

    @Test fun `TYPE_03 は Fizz のみで Buzz なし`() {
        assertEquals("Fizz", FizzBuzzType.TYPE_03.generate(3))
        assertEquals("5", FizzBuzzType.TYPE_03.generate(5))
        assertEquals("FizzBuzz", FizzBuzzType.TYPE_03.generate(15))
    }

    @Test fun `番号 1 は TYPE_01`() = assertEquals(FizzBuzzType.TYPE_01, FizzBuzzType.create(1).getOrNull())
    @Test fun `番号 2 は TYPE_02`() = assertEquals(FizzBuzzType.TYPE_02, FizzBuzzType.create(2).getOrNull())
    @Test fun `番号 3 は TYPE_03`() = assertEquals(FizzBuzzType.TYPE_03, FizzBuzzType.create(3).getOrNull())
    @Test fun `未定義の番号は失敗`() = assertTrue(FizzBuzzType.create(99).isFailure)
}
